src/vs/platform/quickinput/browser/quickPickPin.ts

```diff
@@ -0,0 +1,81 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Codicon } from 'vs/base/common/codicons';
+import { localize } from 'vs/nls';
+import { ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { IQuickPick, IQuickPickItem, IQuickPickItemButtonEvent, QuickPickItem } from 'vs/platform/quickinput/common/quickInput';
+import { IStorageService, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { ThemeIcon } from 'vs/platform/theme/common/themeService';
+
+/**
+ * Initially, adds pin buttons to all @param quickPick items.
+ * When pinned, a copy of the item will be moved to the end of the pinned list and any duplicate within the pinned list will
+ * be removed. Pin and pinned button events trigger updates to the underlying storage.
+ * Shows the quickpick once formatted.
+ */
+export async function showWithPinnedItems(accessor: ServicesAccessor, storageKey: string, quickPick: IQuickPick<IQuickPickItem>): Promise<void> {
+	const storageService = accessor.get(IStorageService);
+	quickPick.onDidTriggerItemButton(async (e) => {
+		if (e.button.iconClass === ThemeIcon.asClassName(Codicon.pin) || e.button.iconClass === ThemeIcon.asClassName(Codicon.pinned)) {
+			quickPick.items = await _formatPinnedItems(storageKey, quickPick, storageService, e);
+		}
+	});
+	quickPick.onDidChangeValue(async value => {
+		// don't show pinned items in the search results
+		quickPick.items = value ? quickPick.items.filter(i => i.type !== 'separator' && !i.buttons?.find(b => b.iconClass === ThemeIcon.asClassName(Codicon.pinned))) : quickPick.items;
+	});
+	quickPick.items = await _formatPinnedItems(storageKey, quickPick, storageService);
+	await quickPick.show();
+}
+
+function _formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, event?: IQuickPickItemButtonEvent<IQuickPickItem>): QuickPickItem[] {
+	const formattedItems: QuickPickItem[] = [];
+	const labels = getPinnedItems(storageKey, storageService).map(item => item.label);
+	const updatedLabels = !!event?.item.label ? updatePinnedItems(storageKey, event.item, storageService, new Set(labels).has(event.item.label)) : labels.filter(l => l !== 'Pinned');
```

Comment:
In your case, `label`s are distinct and can be used as an identifier... but there could be 2 identical labels with different details or description and this will cause some double pinning/unpinnning.

QuickPickItem has a concept of an `id`. I think that you should use that instead and if an item doesn't have an `id` then you should use the combination of `label`, `description`, and `detail` as the id.

Reply:
in this case, the behavior you described is what we want... 

list:
- a
- b
- c

pin a

new list:
- b 
- c 
- a

So when only a `label` is provided, that's the behavior

Those with IDs will allow duplicates inherently.

Sound good?

Reply:
I think an issue with using IDs is duplicates would lose the group information when they go to the pinned list. We could require pinned items to have ids and call that out in the format pinned function jsdoc?

Reply:
This `new Set(labels).has(changedItem.label)` is the thing I'm a bit hung up on...

So here's an example:

```
Pinned:
* Megan - vscode team
* Tyler - vscode team
--------
* Daniel - vscode team
* Daniel - aml team
* Alex - vscode team
```

> Where the text before the `-` is `.label` and the text after the `-` is `.description` on the quick pick item

Since `Daniel - vscode team` and `Daniel - aml team` have the same `.label` value... what happens when:

* Pin `Daniel - vscode team`
* Unpin `Daniel - vscode team`
* Pin `Daniel - aml team`
* Unpin `Daniel - aml team`

Does it still work as expected?